File: Trie/SearchSuggestionsSystem.cpp
struct Node {
    Node* arr[26];

    vector<string> suggestions;

    Node() {
        for (int i = 0; i < 26; i++) {
            arr[i] = nullptr;
        }
    }

    bool valueExists(char c) {
        return (arr[c-'a'] == NULL) ? false : true;
    }

    void setChar(char c) {
        arr[c-'a'] = new Node();
    }

    Node* getChar(char c) {
        return arr[c-'a'];
    }

    void modifySuggestions(string word) {
        if(suggestions.size()<3) {
            suggestions.push_back(word);
        }
    }

    vector<string> getSuggestions() {
        return suggestions;
    }

};

class Trie {
    private:
    Node* root;

    public:
    Trie() {
        root = new Node();
    }

    void insert(string word) {
        Node* crawler = root;

        for(char c: word) {
            if(!crawler->valueExists(c)) {
                crawler->setChar(c);
            }
            crawler = crawler->getChar(c);
            crawler->modifySuggestions(word);
        }
    }

    vector<vector<string>> getSuggestions(string word) {
        vector<vector<string>> ans;
        bool prefixStillValid = true;
        Node* crawler = root;

        for (char c : word) {
            if (!prefixStillValid || !crawler->valueExists(c)) {
                ans.push_back({});
                prefixStillValid = false; 
            } else {
                crawler = crawler->getChar(c);
                ans.push_back(crawler->getSuggestions());
            }
        }

        return ans;
    }
};

class Solution {
public:
    vector<vector<string>> suggestedProducts(vector<string>& products, string searchWord) {
        Trie t;

        sort(products.begin(),products.end());

        for(string s: products) {
            t.insert(s);
        }

        return t.getSuggestions(searchWord);
    }
};

File: Trie/SearchSuggestionsSystem.java
package Trie;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class SearchSuggestionsSystem {
    private class TrieNode {
        TrieNode[] trieArray = new TrieNode[26];
        boolean isEnd = false;
        List<String> mapping = new ArrayList<>();
    }

    private final TrieNode root = new TrieNode();
    
    public void insert(String word) {
        TrieNode node = root;
        for(char c: word.toCharArray()) {
            if(node.trieArray[c-'a'] == null) {
                node.trieArray[c-'a'] = new TrieNode();
            }
            node = node.trieArray[c-'a'];

            if(node.mapping.size() < 3) node.mapping.add(word);
        }
        node.isEnd = true;
    }

    public List<List<String>> searchSuggestions(String searchWord) {
        List<List<String>> ans = new ArrayList<>();

        TrieNode node = root;

        boolean noMatch = false;

        for(char c: searchWord.toCharArray()) {
            if(noMatch || node.trieArray[c-'a'] == null) {
                ans.add(new ArrayList<>());
                noMatch = true;
            }
            else {
                node = node.trieArray[c-'a'];
                ans.add(node.mapping);
            }

        }
        
        return ans;
    }
    

    public List<List<String>> suggestedProducts(String[] products, String searchWord) {
        TrieNode trie = new TrieNode();

        Arrays.sort(products);

        for(String product: products) {
            this.insert(product);
        }

        
        return searchSuggestions(searchWord);
    }
}
